lec07-pop-models.Rmd

```diff
@@ -1,4 +1,617 @@
+derivatives, etc.). **Ordinary** means that the derivatives are full derivatives (as opposed to partial
+derivatives which are written $\partial N / \partial t$). 
+
+This means that the function we'll be using is called `ode`, which stands for 'ordinary differential equation'. 
+
+```{r}
+help(ode)
+# The possible methods are all different ways of numerically solving DEs.
+
+#stiff vs. non-stiff : no precise definition, but 'stiff' generally means it's possible to be numerically unstable if the step size is too big. 
+```
+
+The method used by `ode` to numerically solve differential equations is called `lsoda`, which automatically
+tries to use the best method depending on the equation. This is why people like to use functions like `ode`:
+the result is often more accurate than Euler's method and might also be faster to run. 
+
+```{r}
+# Example: deSolve with logistic eqn
+
+library(deSolve)
+
+# now we need to define our function a bit differently to be in the format that `ode` uses
+logistic_fn <- function(t,state,parameters) { 
+  # Calculates dN/dt for the logistic equation
+  
+  # t: time point at which to evaluate derivative (doesn't actually change anything in this example)
+  # state: vector of variables (here it's just N)
+  # parameters: vector of parameters c(r, K)
+  
+  N <- state 
+  
+  r <- parameters[1] # the first element of the parameters vector is r
+  K <- parameters[2] # the second element of the parameters vector is K
+
+  #rate of change
+  dN = r*N*(1-N/K)
+    
+  #return rate of change
+  list(c(dN))
+}
+
+parameters <- c(r=0.5,K=50)
+state <- c(N=10)
+times <- seq(0, 50, by = 0.01)
+
+help(ode) # look at the documentation to learn about the parameters
+result <- ode(y = state, times = times, func = logistic_fn, parms = parameters)
+```
+
+The output of `ode` is a vector that contains the times we requested and its calculated values of N. 
+
+Notice that by using `ode` we didn't have to write a loop explicitly ourselves, and we didn't have to
+choose a timestep. `ode` does under the hood a version of what we did explicitly before. 
+
+```{r}
+print(head(result,10)) # result has a list of times and values of N at those times
+print(typeof(result)) # result is a 'double': a vector of numbers.
+```
+
+```{r}
+
+result <- data.frame(result) # convert it to a dataframe so we can use ggplot
+
+ggplot(result) +
+  geom_point(aes(x = time, y = N))
+```
+
+Now let's make a nice plot of numerical solutions from a few different starting points.
+
+```{r}
+library(dplyr)
+
+parameters <- c(r=0.5,K=50)
+initial_conditions <- seq(10, 80, by = 10)
+times <- seq(0, 50, by = 0.01)
+
+results <- list() # make an empty list
+count <- 1
+
+for (N0 in initial_conditions) {
+  
+  result <- ode(y = N0, times = times, func = logistic_fn, parms = parameters)
+  result <- data.frame(result)
+  
+  results[[count]] <- result
+  
+  count <- count + 1
+}
+
+all_results <- bind_cols(results)
```

Comment:
I'm going to submit a PR to your fork of some code suggestions for this. It will be way easier than doing it here.

Reply:
PS: submitted a [PR to your fork](https://github.com/mbonsma/rcourse/pull/2). Let me know if you want me to explain any of it. `do` just applies a function to each initial condition in the data.frame, like the loop does, but in the tidyverse style of doing it. So that the output is a data frame, which you can than continue in a pipe for whatever else you want to do.

Reply:
If you merge that PR into your fork, it will update automatically here.